component/componenttest/lifecycle_verifier.go

```diff
@@ -0,0 +1,79 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package componenttest
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config/configmodels"
+)
+
+// GetExtensionConfigFn is used customize the configuration passed to the verification.
+// This is used to change ports or provide values required but not provided by the
+// default configuration.
+type GetExtensionConfigFn func() configmodels.Extension
+
+// VerityExtensionLifecycle is used to test if an extension type can handle the typical
+// lifecycle of a component. The getConfigFn parameter only need to be specified if
+// the test can't be done with the default configuration for the component.
+func VerityExtensionLifecycle(t *testing.T, factory component.ExtensionFactory, getConfigFn GetExtensionConfigFn) {
+	ctx := context.Background()
+	host := NewAssertNoError(t)
+	extCreateParams := component.ExtensionCreateParams{
+		Logger:               zap.NewNop(),
+		ApplicationStartInfo: component.DefaultApplicationStartInfo(),
+	}
+
+	var activeExt, builtExt component.Extension
+	defer func() {
+		// If the shutdown happens here there were already some errors on the test.
+		// Ignore errors on this attempt to clean-up.
+		if activeExt != nil {
+			_ = activeExt.Shutdown(ctx)
+		}
+		if builtExt != nil {
+			_ = builtExt.Shutdown(ctx)
+		}
+	}()
+
+	if getConfigFn == nil {
+		getConfigFn = factory.CreateDefaultConfig
+	}
+
+	for i := 0; i < 3; i++ {
```

Comment:
Nit: twice should be enough?

Reply:
Also logic is too complicated, much simpler to write code twice:

firstExt, err = factory.CreateExtension(ctx, extCreateParams, getConfigFn())
require.NoError(t, err)
require.NoError(t, firstExt.Start(ctx, host))
secondExt, err = factory.CreateExtension(ctx, extCreateParams, getConfigFn())
require.NoError(t, err)
assert.NoError(t, firstExt.Shutdown(ctx))
require.NoError(t, secondExt.Start(ctx, host), "Extension type: %s", factory.Type())
assert.NoError(t, secondExt.Shutdown(ctx))

Reply:
Fixed in https://github.com/open-telemetry/opentelemetry-collector/pull/2679/commits/20812d507c9c40c3fa99896a1db63feaa632ef57